### ISSUE TITLE
Recreating D3 example failed

### ISSUE THREAD

**Issue by Cheukting** [pyodide/pyodide#3060]
## 🐛 Bug

I tried to recreate an example for [this D3 network graph](https://d3-graph-gallery.com/graph/network_basic.html) in PyScript ([minimal example here](https://github.com/Cheukting/Cheukting.github.io/blob/master/assets/html/min-network-d3.html)) then came across this error:
```
Uncaught TypeError: jsargs is not iterable (cannot read property undefined)
    at Proxy.apply (pyproxy.gen.ts:1145:19)
    at Dispatch.call (d3-dispatch.js:58:18)
    at step (d3-force.js:228:13)
    at timerFlush (d3-timer.js:49:15)
    at wake (d3-timer.js:58:5)
```

### To Reproduce

See https://github.com/Cheukting/Cheukting.github.io/blob/master/assets/html/min-network-d3.html

### Expected behavior

See https://d3-graph-gallery.com/graph/network_basic.html

**Comment by rth** [comment]
In Firefox, the error was a bit different,
```
Uncaught TypeError: jsargs is undefined
    apply pyproxy.gen.ts:1145         [pyproxy.gen.ts:1145:39](https://cdn.jsdelivr.net/pyodide/v0.20.0/src/js/pyproxy.gen.ts)
    call d3-dispatch.js:58
    step d3-force.js:228
    timerFlush d3-timer.js:49
```
I had a look but I have no idea what is wrong exactly and whether it's a user code error or an issue in Pyodide. Passing extra arguments into that callback function doesn't really help.

**Comment by leriomaggio** [comment]
Ok so I played with it a little bit and here is what I found.

1. Re error: the error message is indeed a bit different but the essence is the same. 

Expanding a bit about this (with the help of Browser inspector)
in `d3-dispatch.js` (around lines `51-59`) 
```
51 call: function(type, that) {
         [...]
57     for (t = this._[type], i = 0, n = t.length; i < n; ++i)
58         t[i].value.apply(that, args);
59  },
```

`L58` gets evaluated because `this` (the `Dispatch`) looks like:
```
this:  Dispatch
    _: 
        end: Array(1)
            0: {name: '', value: Proxy}
            length: 1
```

`args` will later be the `jargs` in question for `PyProxy` that is indeed `undefined` as **none** of the previous lines in `call` (from `d3-dispatch.js`) where evaluated. 

Therefore, from the **Pyodide** side, these situations could be potentially accounted for by checking whether `jargs` is `undefined` or not and so call the `PyObject` accordingly. 

I don't know whether that makes any sense, tbh (_disclaimer_ 😅 ), but I think the issue here is perhaps a bit more philosophical (re specific example, see below)

2. As for  the specific example:
I sensed the error could've been related to `Proxy` conversion  - that is perhaps not the right thing to do here. 

TLDR; 
I replaced `create_proxy` with `create_once_callable` for the `ticked` PyFunction and it works!
This function is meant to be called once (`on("end"..)` in fact) 
Therefore I thought that a quick to consume `PyProxy` object is what was more appropriate... and it worked 🎉  

More specifically on the issue: I looked carefully at what happens on the `call stack` when it generates the error. 

- With **Porxy**, `d3-dispatch` evaluates immediately `pyproxy-gen-ts` (`L1145`) that raises the exception as the object in `this` is a Proxy (not being available yet from the promise? 🤔 )

- With **Once Callable**, `this` contains directly the `wrapper` function ready to be invoked by `pyodide-asm` (`L14`) which is indeed already the callable we need there :) 

I hope this helps in shading some light on the issue at hand?

**Comment by leriomaggio** [comment]
Further on that example, @Cheukting and I tried to take it a step further. 
In fact, while the example works `on("end", create_once_callable(pyfunc))`, it doesn't if we use `tick` instead.

In fact, changing that to `on("tick", create_once_callable(pyfunc))` fails because the callable object is called more than once, hence is already gone for good after the first call. 

Therefore, in this "loop" series of calls (as in `tick`) I was expecting to use the `create_proxy` to work but apparently we were back to square one. 

The only way I could manage to have it working, was to completely delegate the `ticked` callback to JS directly, invoked by PyScript/Pyodide:

```
simulation = (d3.forceSimulation(nodes_js)
                      .force("link", d3.forceLink().id(id_fn).links(edges_js))
                      .force("charge", d3.forceManyBody().strength(-400))
                      .force("center", d3.forceCenter(width / 2, max_height / 2))
                      .on("tick", js.ticked)
                      )
```

with 
```
<script>
  // This function is run at each iteration of the force algorithm, updating the nodes position.
  function ticked() {
    d3.select('svg').selectAll("line")
            .attr("x1", function (d) {
              return d.source.x;
            })
            .attr("y1", function (d) {
              return d.source.y;
            })
            .attr("x2", function (d) {
              return d.target.x;
            })
            .attr("y2", function (d) {
              return d.target.y;
            });

    d3.select('svg').selectAll("circle")
            .attr("cx", function (d) {
              return d.x + 6;
            })
            .attr("cy", function (d) {
              return d.y - 6;
            });
  }

</script>
```

Sharing this just because I am still quite puzzled on why `JSProxy` is not working there 🤔

**Comment by hoodmane** [comment]
The d3 code that calls this function looks like:
```
call: function(type, that) {
    if ((n = arguments.length - 2) > 0)
      for (var args = new Array(n), i = 0, n, t; i < n; ++i)
        args[i] = arguments[i + 2];
    if (!this._.hasOwnProperty(type))
      throw new Error("unknown type: " + type);
    for (t = this._[type], i = 0, n = t.length; i < n; ++i)
      t[i].value.apply(that, args);
}
```
In the call `arguments.length` is 2, so `args` is left `undefined`. `PyProxy.apply` attempts to spread `args` into `callPyObject` but does nothing to protect against the possibility that `args` is undefined.

**Comment by rth** [comment]
Thanks for investigating @leriomaggio and @hoodmane for fixing it in #3101!
